Second development task elkészült

NemesTom opened the pull request:
parseUnit függvény létrehozva, main már fájlokat vár argumentumként, ebből az új formatnak megfelelő eredményeket ad. Van továbbá 3 unitunk, a GHA tesztet egy bash scripttel csináltuk meg, itt volt pár baki, figyelmetlenség ami a commitok nevéből és népességéből is látszik, de végül rájöttünk a hiba okára, így sikeresen lefut minden. Kérjük nézzétek meg, reméljük jó.

NemesTom committed:
Hitlog remove, fight results are in the new format and printed into a txt file

The results file is named file.txt for the time being

NemesTom committed:
Ads units folder with 3 units inside

NemesTom committed:
Update .gitignore with .json extension

NemesTom committed:
Results file now has the "AttackerAttacked" name format

Also minor, non functional changes to code

NemesTom committed:
Results format correction

NemesTom committed:
Update .gitignore

LilikF committed:
Add parseUnit for Character class

LilikF committed:
Moved file writing function to util

LapatKrisz committed:
Added GHA workflow and updated gitignore

LapatKrisz committed:
Fixed GHA workflow and updated gitignore

NemesTom committed:
Add GHA input test, correct results, modifies gitignore

NemesTom committed:
Merge branch 'second_development_task' of https://github.com/Teaching-projects/SZE-MOSZE-2020-Merida_dualthrust into second_development_task

NemesTom committed:
Correct permission for bash script

NemesTom committed:
Corrects name of shell script in Test.yml

NemesTom committed:
Update Test.yml

NemesTom committed:
GHA yaml modification

NemesTom committed:
Highly experimental GHA yaml modification

Files have been rearranged in a rather ugly way, will be fixed in a future commit.

NemesTom committed:
Syntax error fix in yml

NemesTom committed:
Syntax error fix in yml (?)

NemesTom committed:
Experimental GHA yml modification

NemesTom committed:
Adds linebreak in run_test.sh

NemesTom committed:
Removes unused diffs

NemesTom committed:
Adds linebreak in run_test.sh

LilikF committed:
Print file contents for debugging

LilikF committed:
Add new line to results file

LilikF committed:
Remove echos from bash file

NemesTom committed:
Test result echo

LapatKrisz committed:
Remove write to file function

LapatKrisz committed:
Update README.md

oliverosz commented on .gitignore:
```diff
@@ -1,8 +1,18 @@
 # Ignore everything
-*
-
+/*
+/*/
+/results/*
```
Ez elég fura megoldás. Normál esetben mindent verziókövetünk, néhány kivétellel, és nem fordítva. De tőlem maradhat.

oliverosz commented on main.cpp:
```diff
@@ -11,23 +11,30 @@ using namespace std;
 
+    if (player1 && player2) 
+    {
+        //The fight method is called, this is where all the logic happens
+        player1->fight(player2);
+        
         //After the fight, we delete the objects we've created
-        delete player;
-        delete enemy;
+        delete player1;
+        delete player2;
```
Na ezért veszélyes a manuális memóriakezelés. Ránézésre jónak tűnik, de ha a 2 közül csak az egyiket sikerül létrehozni, akkor az nem lesz felszabadítva. Nyilván a mainben nem okoz nagy gondot egy ilyen hiba, de attól még hibás. Tőlem használhattok dinamikus memóriakezelést, de akkor legyen helyes.

oliverosz commented on main.cpp:
```diff
@@ -11,23 +11,30 @@ using namespace std;
+    {
         //If the input parameters are not correct, we print a helper to the console
-        cout << "Incorrect input parameters. Please use the following pattern: Name1(string) HP1(int) ATK1(int) Name2(string) HP2(int) ATK2(int)" << endl;
+        cout << "Incorrect input parameters. Please check if the input file paths you entered were correct." << endl;
```
Itt meg illene valami nemnulla visszatérési érték.

oliverosz commented on Utility.h:
```diff
@@ -2,12 +2,13 @@
+#include <vector>
 
 class Utility {
 
     public:
         static bool isNumber(std::string);
-
+        static std::vector<std::string> split(std::string, char);
```
Ha csak static metódusai vannak az osztálynak, akkor az inkább egy namespace, helyesebb lenne annak hívni. Mert most megengedi, hogy példányosítsák, aminek meg nincs értelme. De amúgy :+1: a utility függvények külön fájlba szervezéséért, csak szebb lenne erre namespace-t használni.

oliverosz replied:
Csak annyit kell csinálni, hogy `class` helyett `namespace`-t kell írni, és kitörölni a `public:` meg a `static` szavakat.

oliverosz commented on Utility.cpp:
```diff
@@ -2,14 +2,45 @@
 #include <iostream>
 #include <string>
 #include <ctype.h>
+#include <vector>
+#include <fstream>
```
A vector és a string már include-olva van a `Utility.h` által, felesleges ide.
A `ctype.h` az C-s header, C++-ban a `cctype` használandó helyette.

oliverosz commented on Utility.cpp:
```diff
@@ -2,14 +2,45 @@
 
 //Method for checking if a given input is a number or not
-bool Utility::isNumber(std::string input) {
+bool Utility::isNumber(std::string input) 
```
Ne másolgassuk feleslegesen az objektumokat, a paraméter legyen const referencia.

oliverosz commented on Utility.cpp:
```diff
@@ -2,14 +2,45 @@
     if (input.size() == 0) return false;
-    for (char c : input) {
-        if (!isdigit(c)) {
+    for (char c : input) 
```
Itt meg további felesleges másolások minden karakterre. `for (const char& c : input)`

oliverosz commented on Utility.cpp:
```diff
@@ -2,14 +2,45 @@
+
+//Method for splitting a string by specific characters - similar to other programming languages' string.split methods
+//Used for conveniently handling the input JSON files
+std::vector<std::string> Utility::split(std::string s, char splitChar) 
```
Szintén const ref paraméter legyen a string.

oliverosz commented on Utility.cpp:
```diff
@@ -2,14 +2,45 @@
+std::vector<std::string> Utility::split(std::string s, char splitChar) 
+{
+	std::vector<std::string> output;
+	std::string current_value = "";
```
A default konstruktor is üres stringet hoz létre, felesleges az `= ""`.

oliverosz commented on Character.h:
```diff
@@ -12,6 +12,7 @@ class Character {
 
 public:
     Character(const std::string, const int, const int);
+    static Character* parseUnit(const std::string);
     std::string getName() const;
```
Itt is referencia legyen a string paraméter, illetve a getName se másolatot adjon vissza az adattagból, hanem const referenciát.

oliverosz replied:
A 2. fele még nincs meg.

oliverosz commented on Character.cpp:
```diff
@@ -22,39 +25,72 @@
+        //std::cout << enemy->getName() << " -> " << this->getName() << std::endl;
         this->HP = (this->getHP() - enemy->getATK());
-        if (this->HP < 0) { this->HP = 0; }
+        if (this->HP < 0)
+        {
+            this->HP = 0; 
```
Felesleges ez a sok `this`, a tagfüggvényben scope-ban vannak az adattagok és más metódusok.

oliverosz commented on run_test.sh:
```diff
@@ -0,0 +1,19 @@
+
+IFS=$'\n'
+
+./a.out units/Tolvaj.json units/Robin.json >> results.txt
```
Az elsőnél jobb lenne `>`-t használni, hogy ha már létezik a fájl, írja felül. És akkor lokálisan egymás után többször is futtatható.

oliverosz commented on Character.cpp:
```diff
@@ -22,39 +25,72 @@
+            //We read the whole file into a string variable using ifstream and stringstream
+            //We do this because since we'll use the split method anyway, a counter variable holding which row we're currently reading is not needed
+            //This saves us a few addition and divide operations here
+            std::stringstream s;
+            s << f.rdbuf();
+            std::string fileContents = s.str();
```
Így is jó, de ez stringstream nélkül is megoldható: `std::getline(f, fileContents, '\0');`

oliverosz reviewed (changes requested):
A `results_correct`ben lévő különálló kimenetek gondolom már nem kellenek.

A README-ben frissítsétek a működés leírását!

A json parsolás a jelenlegi stádiumban megfelelő, majd úgyis tovább kell fejleszteni a későbbi feladatok során. A hibák inkább a többi részben vannak, akár az előző feladat megoldásában is. De főleg apróságok.

LapatKrisz committed:
Minor changes

-Add recommended not-null return value in main.cpp
-Updated run_test.sh: in case of multiple runs the results.txt gets overwritten
-Updated Utility.cpp: remove and modify includes where needed;
remove the unnecessary =""

LilikF committed:
Safe method for deleting existing objects, remove unnecessary code from previous tasks

LilikF committed:
Replace input parameteres with references where needed, remove unneeded "this->" occurences from Character.cpp

oliverosz commented on Utility.cpp:
```diff
@@ -43,4 +40,11 @@ std::vector<std::string> Utility::split(std::string s, char splitChar)
+//Delete every existing Character object given as input
+void Utility::safeDeleteCharacters(const std::vector<Character*>& characters) {
+	for (Character* c : characters) {
+		if (c) delete c;
```
Nem kell az `if (c)`. A `delete nullptr` nem csinál semmit.

oliverosz reviewed (changes requested):
Bár még nem kértetek újabb reviewt, átnéztem, mi hiányzik még. Kommentekben jeleztem.

LilikF committed:
Remove unneeded if statement from deleteCharacters function

LilikF committed:
Corrected function name in main

LilikF committed:
Create namespace out of Utility class

LilikF committed:
Make character getters return a constant reference

LilikF commented:
@oliverosz Köszönjük, gyorsan ledolgoztam őket, remélem jó lett így. Egyedül a stringstream helyetti std::getline null delimiteres dologgal vagyok meglőve, akárhogy próbálom nem olvas be semmit úgy, és neten se nagyon találtam rá példát. Lehet hogy én szúrok el valamit, ha nagyon fontos akkor beszélek még a srácokkal és megpróbáljuk megoldani. Ha viszont jó a stringstream is, akkor megköszönnénk ha így is átmenne a PR.

oliverosz commented on Character.cpp:
```diff
@@ -10,17 +10,17 @@
     }
 
-    int Character::getHP() const
+    int const & Character::getHP() const
```
Egyszerű típusoknál felesleges referenciát visszaadni, ott a másolás nem hív meg másoló konstruktort, meg csak 4 bájt az egész. De mindegy, maradhat.

LilikF replied:
Áh, értem. Ezt nem tudtam, köszönöm!

oliverosz reviewed (approved):
@LilikF Nem tudom, mi lehet a baj a getline-nal. Én kipróbáltam, és működött. De jó a stringstream is.

@hegyhati merge into master, Master

LilikF commented:
@hegyhati 
@oliverosz feljebb jelezte, hogy jó lesz így a PR. Be tudnád mergelni nekünk a masterbe? Kezdenék dolgozni az új feladaton, de talán szerencsésebb ha nem egy levegőben lógó branchről ágazok el, hanem a masterről (bár különbség nem lesz a kettő között már, de azért mégis). Köszönjük!
